### PR TITLE
Make unix sockets optional

### PR DESCRIPTION
**Pull request by ahedberg**
For issue #5028

### CODE REVIEW

File: src/core/iomgr/tcp_client_posix.c [hunk 1]
```diff
@@ -83,6 +84,15 @@ static int prepare_socket(const struct sockaddr *addr, int fd) {
             strerror(errno));
     goto error;
   }
+#else
+  if (!grpc_set_socket_nonblocking(fd, 1) || !grpc_set_socket_cloexec(fd, 1) ||
```
Comment (ctiller) [inline src/core/iomgr/tcp_client_posix.c]: There's too much duplicated logic in this block.

Can we instead introduce a function: `bool grpc_supports_low_latency(sockaddr *)` that returns true always for the `!GPR_HAVE_UNIX_SOCKET` case, and `addr->sa_family != AF_UNIX` for the `GPR_HAVE_UNIX_SOCKET` case?

File: src/core/iomgr/tcp_server_posix.c [hunk 2]
```diff
@@ -288,6 +297,16 @@ static int prepare_socket(int fd, const struct sockaddr *addr,
             strerror(errno));
     goto error;
   }
+#else
+  if (!grpc_set_socket_nonblocking(fd, 1) || !grpc_set_socket_cloexec(fd, 1) ||
```
Comment (ctiller) [inline src/core/iomgr/tcp_server_posix.c]: Similarly, use `bool grpc_supports_low_latency(sockaddr *)` here.

**Comment by ctiller** [comment]
General comment: we've currently got very few inlined `#ifdef's` across the codebase. I'd like to keep it this way as it helps readability tremendously.

Can we introduce a module (unix_sockets_posix.{h,c}) that defines a small API for the things that we'd want to use unix sockets for? I'd imagine then having a unix_sockets_noop.c variant of the implementation that just stubs these routines out.

**Comment by ahedberg** [comment]
Sure, but keep in mind that I have no idea what uses unix sockets vs. regular posix sockets (or why both are required).

**Comment by ahedberg** [comment]
Module added, PTAL.

**Comment by ahedberg** [comment]
<ping>

File: src/core/iomgr/unix_sockets_posix.c [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  strcpy(un->sun_path, name);
+  addrs->addrs->len = strlen(un->sun_path) + sizeof(un->sun_family) + 1;
+  return addrs;
+}
+
+int grpc_is_unix_socket(const struct sockaddr *addr) {
+  return addr->sa_family == AF_UNIX;
+}
+
+void unlink_if_unix_domain_socket(const struct sockaddr *addr) {
```
Comment (ctiller) [inline src/core/iomgr/unix_sockets_posix.c]: External linkage functions need a grpc_ prefix

File: src/core/iomgr/unix_sockets_posix.c [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+  }
+  struct sockaddr_un *un = (struct sockaddr_un *)addr;
+  struct stat st;
+
+  if (stat(un->sun_path, &st) == 0 && (st.st_mode & S_IFMT) == S_IFSOCK) {
+    unlink(un->sun_path);
+  }
+}
+
+int parse_unix(grpc_uri *uri, struct sockaddr_storage *addr, size_t *len) {
```
Comment (ctiller) [inline src/core/iomgr/unix_sockets_posix.c]: External linkage functions need a grpc_ prefix

File: src/core/iomgr/unix_sockets_posix.c [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  struct sockaddr_un *un = (struct sockaddr_un *)addr;
+
+  un->sun_family = AF_UNIX;
+  strcpy(un->sun_path, uri->path);
+  *len = strlen(un->sun_path) + sizeof(un->sun_family) + 1;
+
+  return 1;
+}
+
+char *unix_get_default_authority(grpc_resolver_factory *factory,
```
Comment (ctiller) [inline src/core/iomgr/unix_sockets_posix.c]: External linkage functions need a grpc_ prefix

**Comment by ctiller** [comment]
Approach looks fine... there are compilation issues that need to be looked at, and some tweaks to naming.

This is looking much better though.

File: src/core/iomgr/unix_sockets_posix.c (hunk 3 above)
Comment (ahedberg) [inline src/core/iomgr/unix_sockets_posix.c]: Done.

File: src/core/iomgr/unix_sockets_posix.c (hunk 4 above)
Comment (ahedberg) [inline src/core/iomgr/unix_sockets_posix.c]: Done.

File: src/core/iomgr/unix_sockets_posix.c (hunk 5 above)
Comment (ahedberg) [inline src/core/iomgr/unix_sockets_posix.c]: Done.

**Comment by ahedberg** [comment]
The compilation errors in the previous run should be fixed now.

**Comment by ctiller** [comment]
LGTM

**Comment by jtattermusch** [comment]
windows failure is a known flake.

**Pull request merged**